File: lib/features/dashboard/domain/entities/storage_info.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'large_file_info.dart';
import 'storage_volume_info.dart';

part 'storage_info.freezed.dart';
part 'storage_info.g.dart'; // Added this back if missing, though it wasn't shown in previous view

@freezed
class StorageInfo with _$StorageInfo {
  // Removed @JsonSerializable on constructor
  const factory StorageInfo({
    required int totalSpace,
    required int freeSpace,
    required int usedSpace,
    required int photosCount,
    required int photosSize,
    required int videosCount,
    required int videosSize,
    required int audioCount, // NEW
    required int audioSize, // NEW
    required int documentsCount, // NEW
    required int documentsSize, // NEW
    required int filesCount,
    required int filesSize,
    required int systemSize,
    required int appsCount, // NEW
    required int duplicateCount, // NEW
    required int duplicateSize, // NEW
    required int potentialSavings, // NEW
    required int similarPhotoCount, // NEW
    required int similarPhotoSize, // NEW
    required List<LargeFileInfo> largeFiles, // NEW
    required List<StorageVolumeInfo> storageVolumes, // NEW
    required int cloudOnlyCount, // NEW
    required int scanDurationMs, // NEW
    required int lastUpdated, // NEW: Timestamp
    required bool isEstimated, // NEW
    required int junkCount, // NEW
    required int junkSize, // NEW
    required int emptyFolderCount, // NEW
    required int apkCount, // NEW
    required int apkSize, // NEW
  }) = _StorageInfo;

  const StorageInfo._();

  /// Empty shell for progressive loading — dashboard shows immediately
  factory StorageInfo.empty() => const StorageInfo(
    totalSpace: 0,
    freeSpace: 0,
    usedSpace: 0,
    photosCount: 0,
    photosSize: 0,
    videosCount: 0,
    videosSize: 0,
    audioCount: 0,
    audioSize: 0,
    documentsCount: 0,
    documentsSize: 0,
    filesCount: 0,
    filesSize: 0,
    systemSize: 0,
    appsCount: 0,
    duplicateCount: 0,
    duplicateSize: 0,
    potentialSavings: 0,
    similarPhotoCount: 0,
    similarPhotoSize: 0,
    largeFiles: [],
    storageVolumes: [],
    cloudOnlyCount: 0,
    scanDurationMs: 0,
    lastUpdated: 0,
    isEstimated: true,
    junkCount: 0,
    junkSize: 0,
    emptyFolderCount: 0,
    apkCount: 0,
    apkSize: 0,
  );

  factory StorageInfo.fromJson(Map<String, dynamic> json) =>
      _$StorageInfoFromJson(json);

  double get usedPercentage => totalSpace == 0 ? 0 : usedSpace / totalSpace;

  int get largeFilesTotalSize => largeFiles.fold(
    0,
    (previousValue, element) => previousValue + element.size,
  );
}

File: android/app/src/main/kotlin/com/movixalabs/cleanmaster/models/StorageInfo.kt
package com.movixalabs.cleanmaster.models

data class StorageInfo(
    // Disk space
    val totalSpace: Long,
    val freeSpace: Long,
    val usedSpace: Long,
    
    // Media counts & sizes
    val photosCount: Int,
    val photosSize: Long,
    val videosCount: Int,
    val videosSize: Long,
    val audioCount: Int,
    val audioSize: Long,
    val documentsCount: Int,
    val documentsSize: Long,
    
    // Derived
    val filesCount: Int,
    val filesSize: Long,
    val systemSize: Long,
    val appsCount: Int,
    
    // Metadata
    val lastUpdated: Long,
    val isEstimated: Boolean,
    
    // NEW: Enhanced features
    val duplicateCount: Int,
    val duplicateSize: Long,
    val potentialSavings: Long,
    val similarPhotoCount: Int,
    val similarPhotoSize: Long,
    val largeFiles: List<LargeFileInfo>,
    val storageVolumes: List<StorageVolumeInfo>,
    val cloudOnlyCount: Int,
    val scanDurationMs: Long,
    // NEW: Junk stats
    val junkCount: Int,
    val junkSize: Long,
    val emptyFolderCount: Int,
    val apkCount: Int, // NEW
    val apkSize: Long // NEW
) {
    fun toMap(): Map<String, Any> = mapOf(
        "totalSpace" to totalSpace,
        "freeSpace" to freeSpace,
        "usedSpace" to usedSpace,
        "photosCount" to photosCount,
        "photosSize" to photosSize,
        "videosCount" to videosCount,
        "videosSize" to videosSize,
        "audioCount" to audioCount,
        "audioSize" to audioSize,
        "documentsCount" to documentsCount,
        "documentsSize" to documentsSize,
        "filesCount" to filesCount,
        "filesSize" to filesSize,
        "systemSize" to systemSize,
        "appsCount" to appsCount,
        "lastUpdated" to lastUpdated,
        "isEstimated" to isEstimated,
        "duplicateCount" to duplicateCount,
        "duplicateSize" to duplicateSize,
        "potentialSavings" to potentialSavings,
        "similarPhotoCount" to similarPhotoCount,
        "similarPhotoSize" to similarPhotoSize,
        "largeFiles" to largeFiles.map { it.toMap() },
        "storageVolumes" to storageVolumes.map { it.toMap() },
        "cloudOnlyCount" to cloudOnlyCount,
        "scanDurationMs" to scanDurationMs,
        "junkCount" to junkCount,
        "junkSize" to junkSize,
        "emptyFolderCount" to emptyFolderCount,
        "apkCount" to apkCount,
        "apkSize" to apkSize
    )
}
